tests/integration/nm/vlan_test.py

```diff
@@ -0,0 +1,83 @@
+#
+# Copyright 2018 Red Hat, Inc.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 2 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+#
+
+from contextlib import contextmanager
+
+from libnmstate import nm
+
+
+ETH1 = 'eth1'
+
+
+def test_create_and_remove_vlan():
+    vlan_desired_state = {'vlan': {'id': 101, 'base-iface': ETH1}}
+
+    with _vlan_interface(vlan_desired_state):
+
+        _assert_vlan_state(vlan_desired_state)
+
+
+@contextmanager
+def _vlan_interface(state):
+    try:
+        _create_vlan(state)
+        yield
+    finally:
+        _delete_vlan(_get_vlan_ifname(state))
+
+
+def _assert_vlan_state(vlan_desired_state):
+    ifname = _get_vlan_ifname(vlan_desired_state)
+    nmdev = nm.device.get_device_by_name(ifname)
+    vlan_state = nm.vlan.get_info(nmdev)
+    assert vlan_desired_state == vlan_state
+
+
+def _create_vlan(vlan_desired_state):
+    ifname = _get_vlan_ifname(vlan_desired_state)
+    con_setting = nm.connection.create_setting(
+        con_name=ifname,
+        iface_name=ifname,
+        iface_type=nm.nmclient.NM.SETTING_VLAN_SETTING_NAME,
+    )
+    vlan_setting = nm.vlan.create_setting(vlan_desired_state,
+                                          base_con_profile=None)
+    with _mainloop():
+        con_profile = nm.connection.create_profile((con_setting, vlan_setting))
+        nm.connection.add_profile(con_profile, save_to_disk=False)
+    with _mainloop():
+        remote_profile = nm.nmclient.client().get_connection_by_id(ifname)
+        nm.device.activate(dev=None, connection=remote_profile)
+
+
+def _delete_vlan(devname):
+    nmdev = nm.device.get_device_by_name(devname)
+    with _mainloop():
+        nm.device.deactivate(nmdev)
+        nm.device.delete(nmdev)
+
+
+@contextmanager
+def _mainloop():
```

Comment:
Isn't there some other test that calls the mainloop as well? How about having some generic shared code for all the things that are not specific for this test?